### ISSUE TITLE
[BUG] folium.GeoJson() with style_function changes input data

### ISSUE THREAD

**Issue by okomarov** [python-visualization/folium#897]
#### Please add a code sample or a nbviewer link, copy-pastable if possible

Example data:
```python

pistes = {"features":[{'geometry': {'coordinates': [[13.7750875, 47.3894653],
     [13.7745458, 47.3897487],
     [13.7744245, 47.3899024],
     [13.7743022, 47.3901266],
     [13.7742754, 47.3904393],
     [13.7745857, 47.3913728],
     [13.7747696, 47.3922164]],
    'type': 'LineString'},
   'properties': {'difficulty': 'easy'},
   'type': 'Feature'},
  {'geometry': {'coordinates': [[13.7748277, 47.3951473],
     [13.7745382, 47.3955505],
     [13.7740868, 47.3961718],
     [13.7738771, 47.3965048],
     [13.7739069, 47.3967489],
     [13.77409, 47.3970407],
     [13.7744923, 47.3973698],
     [13.7748859, 47.3977331],
     [13.7753385, 47.3980947],
     [13.7756779, 47.3983777],
     [13.7758294, 47.3985963],
     [13.7758483, 47.3987251],
     [13.7758303, 47.3988537]],
    'type': 'LineString'},
   'properties': {'difficulty': 'advanced'},
   'type': 'Feature'},
  {'geometry': {'coordinates': [[13.7750875, 47.3894653],
     [13.7752798, 47.38999],
     [13.7753808, 47.3905846],
     [13.7754171, 47.391452],
     [13.7754382, 47.3919579],
     [13.7751964, 47.3927863],
     [13.7759708, 47.3937253],
     [13.7764045, 47.3944936],
     [13.7766345, 47.3947044],
     [13.7774691, 47.3955802],
     [13.7775424, 47.3958695],
     [13.7773201, 47.3963544],
     [13.7770295, 47.3965058],
     [13.777536, 47.3973757]],
    'type': 'LineString'},
   'properties': {'difficulty': 'intermediate'},
   'type': 'Feature'}]}
```

Example:

```python

m = folium.Map(location=[47.4, 13.78],tiles='OpenStreetMap',zoom_start=14)

diff_to_colour = {'easy':'blue', 'intermediate':'red','advanced':'black'}
def piste_style_function(feature):
    return {'opacity': 1, 
            'weight': 1, 
            'color': diff_to_colour[feature['properties']['difficulty']]}

folium.GeoJson(pistes, style_function=piste_style_function).add_to(m)
m
```

Then, in a separate notebook cell (after map rendered):
```python
pistes

{'features': [{'geometry': {'coordinates': [[13.7750875, 47.3894653],
     [13.7745458, 47.3897487],
     [13.7744245, 47.3899024],
     [13.7743022, 47.3901266],
     [13.7742754, 47.3904393],
     [13.7745857, 47.3913728],
     [13.7747696, 47.3922164]],
    'type': 'LineString'},
   'properties': {'difficulty': 'easy',
    'highlight': {},
    'style': {'color': 'blue', 'opacity': 1, 'weight': 1}},
   'type': 'Feature'},
  {'geometry': {'coordinat 
...
```

Notice the additional `style` property.

#### Problem description
The style function should **NOT** edit the data in place. This become problematic in the following cases:
* You want to re-use a pre-filtered collection of features
* You want to use several layers (where you tweak styles)

#### Expected Output
Unchanged `pistes`

#### Output of ``folium.__version__``
'0.5.0'

**Comment by jtbaker** [comment]
It is a little clunky and makes the datasource larger, but I think this adding this style literal dictionary/object as a property for each feature is the most straightforward way to implement this in Python. It allows us to use Leaflet's `setStyle` method to call a function and loop through the set of features to set each feature's style directly using the `style` key in each feature's `properties`, which it maps to nicely without any additional parsing. You can see where this happens in the `GeoJson` template [here](https://github.com/python-visualization/folium/blob/master/folium/features.py#L382).

Keeping this part of the transformation on the Python side allows us to have more direct control over the data directly vs writing a JS function like `getColor` (line 26) and `style` (line 37) like in [this fiddle](https://jsfiddle.net/jaseface87/o67xc00h/26/) could be challenging to do in a Jinja2 template that can flexibly handle a variety of inputs. It seems like would have to be pretty verbose, since it would involve writing a template JS function to call other JS functions to evaluate for the properties - how `style` calls `getColor` in the fiddle.

I only see the current situation as problematic if you already have properties named `style` or `highlight` in your GeoJson feature properties that are being overwritten in the `self.data` update, or if the size of the data is absolutely critical.

I'm still learning JS and definitely not an expert, however. Do you have any ideas on how to implement this in folium?

**Comment by okomarov** [comment]
@jtbaker Thanks for the answer. 

I feel I should clarify the side-effect that is caused by this behavior. In my use case I color pistes with red and blue. The same piste can change color on different days. So, I keep the red pistes and blue pistes on two separate layers (so I can toggle it on/off). However, by writing into the same common data, only the latest style is plotted. 

One workaround is to force local copies of the same data with e.g. `folium.GeoJson('filepath', overlay=True)` for each plotted layer. On the other hand, I thought each layer is stored in the map object and it would be a matter of copying data for each layer. I understand this might cause data duplication, hence a better case would be to store the style in each layer. 

I am happy to look into this with some direction, unless it's not worth the effort and the  `folium.GeoJson('filepath',...` is good enough.

**Comment by ocefpaf** [comment]
@okomarov I did not looked into this enough but it seems that copying the data before applying the style may fix this, no?

Do you want to try a PR?

**Comment by okomarov** [comment]
@ocefpaf I used the workaround mentioned above, so probably not worth unless others experience similar problems

**Comment by ocefpaf** [comment]
OK. Let's close this then. Thanks for the feedback!

**Comment by arkanoid87** [comment]
been wasting hours on this and still I've not solved the problem

I'm adding more layers using a loop function over the target_col variable
```
            style_function=lambda feature: {
                'fillOpacity': 0.5,
                'weight': 0.5,
                'color': "#8b0000" if feature['properties'][target_col] > -1 else "#808080",
                'fillColor': colorscale(feature['properties'][target_col]) if feature['properties'][target_col] > -1 else "#d4d4d4"
            },
```

I always get `TypeError: '>' not supported between instances of 'NoneType' and 'int'` on m.save(...) on the second feature of the second layer added.

Everything works if I loop over the target_col only once, it always returns error on second layer even if I set global constant geodataframe in input, even when passing a fresh copied database in input.

Very frustrating

**Comment by Conengmo** [comment]
Hi @arkanoid87 , this issue you are replying to is no longer relevant, we no longer change the input data in recent folium versions. I suggest you double check you input data. If you really think the issue lies with folium, please open a new ticket and provide a minimal, self-contained example that will help us debug the issue.
